crates/flipperzero/src/filesystem.rs

```diff
@@ -0,0 +1,301 @@
+            sys::FS_Error_FSE_INVALID_PARAMETER => Self::InvalidParameter,
+            sys::FS_Error_FSE_DENIED => Self::Denied,
+            sys::FS_Error_FSE_INVALID_NAME => Self::InvalidName,
+            sys::FS_Error_FSE_INTERNAL => Self::Internal,
+            sys::FS_Error_FSE_NOT_IMPLEMENTED => Self::NotImplemented,
+            sys::FS_Error_FSE_ALREADY_OPEN => Self::AlreadyOpen,
+            _ => unimplemented!(),
+        }
+    }
+}
+
+/// Trait comparable to `std::Read` for the Flipper stream API
+pub trait Read {
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize, Error>;
+}
+
+/// Enumeration of possible methods to seek within an I/O object.
+///
+/// It is used by the Seek trait.
+pub enum SeekFrom {
+    Start(i32),
+    End(i32),
+    Current(i32),
+}
+
+/// Trait comparable to `std::Seek` for the Flipper stream API
+pub trait Seek {
+    fn seek(&mut self, pos: SeekFrom) -> Result<usize, Error>;
+
+    fn rewind(&mut self) -> Result<(), Error> {
+        self.seek(SeekFrom::Start(0))?;
+        Ok(())
+    }
+
+    fn stream_len(&mut self) -> Result<usize, Error> {
+        let old_pos = self.stream_position()?;
+        let len = self.seek(SeekFrom::End(0))?;
+
+        // Avoid seeking a third time when we were already at the end of the
+        // stream. The branch is usually way cheaper than a seek operation.
+        if old_pos != len {
+            self.seek(SeekFrom::Start(old_pos as i32))?;
+        }
+
+        Ok(len)
+    }
+
+    fn stream_position(&mut self) -> Result<usize, Error> {
+        self.seek(SeekFrom::Current(0))
+    }
+}
+
+/// Trait comparable to `std::Write` for the Flipper stream API
+pub trait Write {
+    fn write(&mut self, buf: &[u8]) -> Result<usize, Error>;
+    fn flush(&mut self) -> Result<(), Error>;
+
+    fn write_all(&mut self, mut buf: &[u8]) -> Result<(), Error> {
+        while !buf.is_empty() {
+            match self.write(buf) {
+                Ok(0) => {
+                    // TODO
+                }
+                Ok(n) => buf = &buf[n..],
+                Err(e) => return Err(e),
+            }
+        }
+        Ok(())
+    }
+}
+
+pub struct OpenOptions(u8, u8);
+
+impl OpenOptions {
+    pub fn new() -> Self {
+        Self(0, 0)
+    }
+
+    /// Read access
+    pub fn read(self, set: bool) -> Self {
+        OpenOptions(
+            if set {
+                self.0 | sys::FS_AccessMode_FSAM_READ
+            } else {
+                self.0 & !sys::FS_AccessMode_FSAM_READ
+            },
+            self.1,
+        )
+    }
+
+    /// Write access
+    pub fn write(self, set: bool) -> Self {
+        OpenOptions(
+            if set {
+                self.0 | sys::FS_AccessMode_FSAM_WRITE
+            } else {
+                self.0 & !sys::FS_AccessMode_FSAM_WRITE
+            },
+            self.1,
+        )
+    }
+
+    /// Open file, fail if file doesn't exist
+    pub fn open_existing(self, set: bool) -> Self {
```

Comment:
To be sure I understand, you're saying that the module should prevent nonsensical states like `FSOM_CREATE_NEW | FSOM_CREATE_ALWAYS` by enforcing mutual exclusivity? Like how `create_new` overrides any previous calls to `create` or `truncate` in `std::fs::OpenOptions`?

(If that's the case, would we want to choose the first one that's set, or the last one that's set?)

Reply:
Exactly. If there's nonsensical values, the best thing we can do is have an API where it's impossible to create those nonsensical values.

We'd want the first one that matches in the list above, since it goes from most specialised (`create_new`) to least (no `OpenOptions` set). For example, if `create_new` is set, then it doesn't really matter what the values of `truncate`, `append` or `create` are—It's going to create a file, which is technically already truncated and you're appending since there can't possibly be any existing data because we're creating a new one.